---
title: "SMCHD1_Sphere_V3"
author: "Raunak Vijayakar"
date: "dd/mm/yyyy"
output: 
  html_document:
    toc: TRUE
    toc_float: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


### Raw counts and annotation

```{r}
library(BiocManager)
library(biomaRt)
library(DESeq2)
library(GenomicFeatures)

sm = read.delim("SMCHD1_RSEMcounts.txt", sep = "")

#select just sphere samples
sm = sm[,c(4:6,10:12)]


#exclude any rows where no counts were found for any condition
sm = sm[rowSums(sm) > 0, ]

#take a look at the reads for each condition
barplot(colSums(sm), ylab="Number of aligned reads", las=2, cex.names=0.8) #!!!!!!
colSums(sm)
# SMCHD1_MZ_4_Sphere_1 is lower than its other conditions by a factor of around 2. Did they find this in their analysis?


#get matching annotations
dr.gtf.db <- makeTxDbFromGFF("Danio_rerio.GRCz10.89.chr.gtf", format="gtf")
ensembl.genes = genes(dr.gtf.db)
fish = useEnsembl(biomart="ENSEMBL_MART_ENSEMBL", host = "asia.ensembl.org", dataset="drerio_gene_ensembl", version = "89") #change host as needed
head(listAttributes(fish))#annotations we can request

#91 doesn't have entrezgene id

bm.annotations = getBM(attributes=c("ensembl_gene_id", "gene_biotype", "description", "zfin_id_symbol", "entrezgene_id"), mart=fish, filters="ensembl_gene_id", values=ensembl.genes$gene_id, uniqueRows=TRUE) #error here sometimes connection issue??

ensembl.genes$zfin_id_symbol = bm.annotations$zfin_id_symbol[ match(ensembl.genes$gene_id, bm.annotations$ensembl_gene_id) ]
ensembl.genes$gene_biotype = bm.annotations$gene_biotype[ match(ensembl.genes$gene_id, bm.annotations$ensembl_gene_id) ]
ensembl.genes$status = bm.annotations$status[ match(ensembl.genes$gene_id, bm.annotations$ensembl_gene_id) ]
ensembl.genes$description = bm.annotations$description[ match(ensembl.genes$gene_id, bm.annotations$ensembl_gene_id) ]
ensembl.genes$entrezgene = bm.annotations$entrezgene[ match(ensembl.genes$gene_id, bm.annotations$ensembl_gene_id) ]


```

``` {r}
#Selecting ONLY gene data that correspond to the annotated genes:

#checking for a hox gene in their data
sm[row.names(sm) == "ENSDARG00000000175", ]
rowMeans(sm[row.names(sm) == "ENSDARG00000000175", ])




nrow(sm)
nrow(sm) == length(row.names(sm) %in% ensembl.genes$gene_id)
sm = sm[row.names(sm) %in% ensembl.genes$gene_id, ]
nrow(sm)



```

365 expressed genes are not found in the 89 ensembl annotation.


### DDS and Filtering

```{r}
#see how many reads are in rRNA/mtRNA

rrna = names(ensembl.genes[ensembl.genes$gene_biotype %in% c("Mt_rRNA", "rRNA")])

unique(ensembl.genes$gene_biotype)
length(names(ensembl.genes[ensembl.genes$gene_biotype %in% "rRNA"]))
names(ensembl.genes[ensembl.genes$gene_biotype %in% "Mt_rRNA"]) #and why is this only 2?? Maybe it's always this low?

#this is a known issue, where the annotation's rrna genes are not found in the actual data
total.rrna = colSums(sm[row.names(sm) %in% rrna,])
sum(row.names(sm) %in% rrna) #WHy is this 0??
#One explanation is that I guessed the wrong annotation, so ensembl.genes does not accord with the content of smchd1 counts. 


barplot(100 * (total.rrna / colSums(sm)), ylab="%rRNA/mtRNA", las=2, cex.names=0.8)

#how many reads are in mitochondrial chromosome?
chrM = names(ensembl.genes[ensembl.genes@seqnames %in% "MT", ])
total.mchrom = colSums(sm[row.names(sm) %in% chrM,])
barplot(100 * (total.mchrom / colSums(sm)), ylab="%MT", las=2, cex.names=0.8)


#remove rRNA and mtRNA
sm = sm[!(row.names(sm) %in% rrna),]
# #remove mitochondrial chromosome reads
# sm = sm[!(row.names(sm) %in% chrM),]


met = read.csv("Sphere_metadata.csv") #sample ids and conditions. no batch indicator, so we won't be controlling for that.

#DESeqDataSet needs countData to be non-negative integers. hence round()
dds = DESeqDataSetFromMatrix(round(sm), met, ~ condition)

dds

dds <- estimateSizeFactors(dds)
dds <- estimateDispersions(dds) #this helps us deal w/fact that we don't have 30 samples/condition

#normalize for read depth
head(counts(dds))
head(counts(dds, normalized=TRUE)) #can do this because we called sizefactors b/f, these number can be compared directly, across condition, not across gene though

hoxb2a = dds[rownames(dds) == "ENSDARG00000000175", ] #not present in sphere v2, present in 4cell v2

hoxd9a = dds[rownames(dds) == "ENSDARG00000059274", ] #present in both v2


#only choosing genes that have a normalized mean expression across conditions >= 10. This removes lowly expressed genes that are hard to differentiate from noise. could also be done after examining p-value histogram of a significance test, comparing pre- and post-filtering.
filter = apply(counts(dds, normalized=TRUE), 1, function(x){ mean(x) > 10 }) #here is the difference between v2 and v3
dds = dds[filter, ]

#filter out only if the average in both conditions is below 20, rather than if the average in either condition is below 20

filter[names(filter) == "ENSDARG00000000175"] #checking for hoxb2a in filter
dds[rownames(dds) == "ENSDARG00000000175", ] #checking for hoxb2a in dds after filter


#apply regularized log transform. 
rld <- rlog(dds)



```

Check for rRNA, and mt rRNA. Ideally these reads will be filtered out during the sequencing process (using ribozero, etc.). The barplots show that the sequencing is already well depleted for both of these sources of noise, as the percentages for both are low. Regardless, remove any genes annotated to be either rRNA, or mtRNA. Filter out genes that are lowly expressed (mean expression < 10), and apply a regularized log transform for PCA, normalizing for library size and putting the count data on a log2 scale.


### PCA
``` {r}
#do PCA
plotPCA(rld, intgroup = c( "condition"))
plotPCA(rld, intgroup = c( "sample_id"))


```

Evidence of possible batch effect going on, but not bad overall.


### Differential expression analysis
```{r}
#all_cond
#since condition is the only contrast, reduced model won't be controlling for sex, batch, etc.
nb = nbinomLRT(dds, full = design(dds), reduced = ~1)
# all_cond = results(nb)
# hist(all_cond$pvalue)
# #though the listed foldchange is calculated for some pair of conditions, the p-value from nbinomlrt is for all conditions.
# 
# sum(is.na(all_cond$padj)) #39 genes have NA p-values and padj
# 
# ###explanation from DESeq2 vignette:###
# # If a row contains a sample with an extreme count outlier then the p value and adjusted p value will be set to NA. These outlier counts are detected by Cookâ€™s distance.
# 
# #For now I will remove these genes and move on, but they might deserve another look
# all_cond = all_cond[!is.na(all_cond$padj),]
# 
# #significance level of 0.05 for FDR. this is a cutoff for the expected false positive rate (proportion of DEGs that were false to total DEGs).
# sum(all_cond$padj < 0.05 & abs(all_cond$log2FoldChange) > 1)
# #FC = 2 ^ log2FC. XUE used 2 fold change cutoff

#nb[rownames(nb) == "ENSDARG00000000175", ] #checking for hoxb2a in nb after filter


#pairwise
all_cond = results(nb, contrast=c("condition", "SMCHD1_Sphere", "WT_Sphere"), independentFiltering = TRUE, alpha=0.05, test="Wald")
hist(all_cond$pvalue) #check this to ensure there isn't batch effect, and that we filtered out the low read genes. Though we can't control for batch in this case.
sum(is.na(all_cond$padj))


#all_cond[rownames(all_cond) == "ENSDARG00000000175", ] #checking for hoxb2a in nb after filter


# nas = all_cond[is.na(all_cond$padj),]
# nas.ann = ensembl.genes[ensembl.genes$gene_id %in% row.names(nas), ]
# nas.ann$zfin_id_symbol

all_cond = all_cond[!is.na(all_cond$padj),] #remove NAs
sum(all_cond$padj < 0.05 & abs(all_cond$log2FoldChange) > 1)
#sum(all_cond$padj < 0.05 & abs(all_cond$log2FoldChange) > 1 & all_cond$baseMean > 20)





# 
# #SMC_vs_DWT
# SMC_vs_DWT = results(nb, contrast=c("condition", "Smchd1", "DWT"), independentFiltering = TRUE, alpha=0.1, test="Wald")
# hist(SMC_vs_DWT$pvalue)
# SMC_vs_DWT = SMC_vs_DWT[!is.na(SMC_vs_DWT$padj),] #remove NAs
# sum(SMC_vs_DWT$padj < 0.01)
# 
# 
# #SMC_vs_DM1
# SMC_vs_DM1 = results(nb, contrast=c("condition", "Smchd1", "DMUT1"), independentFiltering = TRUE, alpha=0.1, test="Wald")
# hist(SMC_vs_DM1$pvalue)
# SMC_vs_DM1 = SMC_vs_DM1[!is.na(SMC_vs_DM1$padj),] #remove NAs
# sum(SMC_vs_DM1$padj < 0.01)

```


To compare expression across all conditions, use the negative binomial likelihood ratio test (nbinomLRT). This is analogous to ANOVA, and allows the comparison of more than two groups. For the pairwise comparison, a Wald test is used, though nbinomLRT would also be acceptable. It is easier here to use Wald, as it means you do not need to subset `dds`. Check for genes containing `NA` in `padj`. For all tests, set the FDR cutoff for significance at 0.05 and |l2fc| > 1, since there are many genes being tested. Across all conditions, there are **1230** significant DEGs, when applying `all_cond$padj < 0.05 & abs(all_cond$log2FoldChange) > 1`.

#### Writing DEG table [NOT RUN]
```{r, eval=F}
#these are ALL degs, NOT just significant
tmp = as.data.frame(all_cond)
tmp = tmp[order(row.names(tmp), decreasing = F),]
ann = as.data.frame(ensembl.genes[ensembl.genes$gene_id %in% row.names(tmp),1:5])
tmp2 = cbind(ann, tmp)
write.csv(tmp2, "Smchd1_Sphere_vs_WT.csv")



```


#### Examine SMCHD1 expression
```{r}
all_cond[row.names(all_cond) == "ENSDARG00000104374",] 

```
The log2foldchange is negative and significant, which indicates reduced expression in the mutant relative to wildtype. Well and truly knocked out at the transcriptional level.

### MA and volcano plots
```{r}
library(scales)

#all_cond
plot(all_cond$log2FoldChange, -log10(all_cond$padj), xlim = c(-10, 10), col = alpha("dodgerblue",0.5), main = "Volcano, SMCHD1 Sphere vs WT Sphere")
abline(h = 10^0.05, v = c(-1,1)) #p = 0.01 and lfc = +/- 2

plotMA(all_cond, ylim=c(-15,15), main = "MA Plot, SMCHD1 Sphere vs WT Sphere")
# 
# #SMC_vs_DWT
# plot(SMC_vs_DWT$log2FoldChange, -log10(SMC_vs_DWT$padj), xlim = c(-15, 15), col = alpha("seagreen",0.5), main = "Volcano, Smchd1 vs DWT")
# abline(h = 2, v = c(-2,2))
# 
# plotMA(SMC_vs_DWT, ylim=c(-15,15), main = "MA Plot, Smchd1 vs DWT")
# 
# #SMC_vs_DM1
# plot(SMC_vs_DM1$log2FoldChange, -log10(SMC_vs_DM1$padj), xlim = c(-15, 15), col = alpha("firebrick",0.5), main = "Volcano, Smchd1 vs DMUT1")
# abline(h = 2, v = c(-2,2))
# 
# plotMA(SMC_vs_DM1, ylim=c(-15,15), main = "MA Plot, Smchd1 vs DMUT1")

```

Produce MA and volcano plots for the pairwise comparisons. MA plots show the difference in gene expression (log fold change) between two conditions, relative to the mean expression across those conditions. Each point represents a particular gene. A gene with a similar expression level in both conditions will be near y = 0. Being further from y = 0 indicates up/down regulation. To compare along the x-axis: For a typically lowly expressed gene (low mean count), there is lower statistical power for an observed expression difference than a highly expressed gene (high A). MA plots tend to have even dispersion relative to the y-axis, which then decreases as x increases. Grey points indicate non-significant data points.

Volcano plots compare adjusted p-value and log fold change. Here I use -log10(adjusted p-value) and log2(fold change). Again, each point represents a gene. The higher on the y-axis, the smaller the adjusted p-value. The more extreme the value on the x-axis, the greater the difference in expression levels across conditions. A greater dispersion means that the two groups have greater differences in gene expression.


#### Identifying specific genes from volcano plots
```{r}
library(ggrepel)

#DM1_vs_DWT
all_cond_d = as.data.frame(all_cond)
all_cond_d$name = row.names(all_cond_d)

ggplot(all_cond_d, aes(x = log2FoldChange, y = -log10(padj), label=name)) + geom_point(alpha = 0.5, color = "dodgerblue") + labs(title = "Volcano, SMCHD1 Sphere vs WT Sphere") + geom_text_repel(data = subset(all_cond_d, -log10(padj) > 30), size = 3)
#setting padj really high, but too many point otherwise


# #SMC_vs_DWT
# SMC_vs_DWT_d = as.data.frame(SMC_vs_DWT)
# SMC_vs_DWT_d$name = row.names(SMC_vs_DWT_d)
# 
# ggplot(SMC_vs_DWT_d, aes(x = log2FoldChange, y = -log10(padj), label=name)) + geom_point(alpha = 0.5, color = "seagreen") + labs(title = "Volcano, Smchd1 vs DWT") + geom_text_repel(data = subset(SMC_vs_DWT_d, -log10(padj) > 30), size = 3)
# 
# #SMC_vs_DM1
# SMC_vs_DM1_d = as.data.frame(SMC_vs_DM1)
# SMC_vs_DM1_d$name = row.names(SMC_vs_DM1_d)
# 
# ggplot(SMC_vs_DM1_d, aes(x = log2FoldChange, y = -log10(padj), label=name)) + geom_point(alpha = 0.5, color = "firebrick") + labs(title = "Volcano, Smchd1 vs DMUT1") + geom_text_repel(data = subset(SMC_vs_DM1_d, -log10(padj) > 40), size = 3)


```

The conditions for gene labelling can be adjusted by applying different logic in `subset()`. Warnings of the form: `Removed k rows containing missing values (geom_point)` are just indicators that k of the data points are outside either `xlim()` or `ylim()`.


### Z-scoring fold changes
```{r}
library(pheatmap)
sig_results = all_cond[all_cond$padj < 0.05 & abs(all_cond$log2FoldChange) > 1,] #the significantly differentially expressed genes from earlier

#transform lfcs then z-score
rlg = rlog(nb)
rlg_sig = assay(rlg)[rownames(sig_results),]
rlg_z = t(apply(rlg_sig, 1, function(x){ (x - mean(x)) / sd(x)})) #normalize by z-scoring

thr = 2
sum(rlg_z > thr)
sum(rlg_z < -thr)
#little data lost from thresholding, but always check! 
#These can also be genes to investigate separately, due to their large scores

rlg_z[rlg_z > thr] = thr
rlg_z[rlg_z < -thr] = -thr
```


### K-means clustering
``` {r}
#determine number of clusters
library(cluster)

#use wss to choose number of clusters
set.seed(123)
k.max <- 15
#1 cluster will almost never be the answer, so we'll leave it out for better visual discrimination
wss <- sapply(2:k.max, 
              function(k){kmeans(rlg_z, k, nstart=10,iter.max = 15 )$tot.withinss})

plot(2:k.max, wss,
     type="b", pch = 19, frame = FALSE, 
     xlab="Number of clusters K",
     ylab="Total within-clusters sum of squares", main = "WSS")
#from wss, I think 5 would be a good value for nclust, based on the elbow method.

#silhouette width to select clusters
SIL = sapply(2:k.max, function(i) {
  results.coef.kmeans =  kmeans(rlg_z, i, nstart=50, iter.max=50)
    mean(silhouette(results.coef.kmeans$cluster, dist(rlg_z))[,3]) })
plot(c(0,SIL), type="b", main = "Silhouette Width")
#silhouette with says use 2 clusters, which is just unreasonable for the data. Small drop from 5 to 6


#in this case it is worth performing hierarchical clustering. 
dd = dist(rlg_z)
clusters.hclust = hclust(dd) #you can choose the agglomeration method here, which would change the end results. 

color = c(colorRampPalette(c("dodgerblue2", "white"))(14), colorRampPalette(c("white", "firebrick2"))(14))
breaksList = seq(-thr, thr, length.out = 29)
heat.map <- pheatmap(rlg_z, cluster_col=FALSE, breaks=breaksList, cluster_rows=clusters.hclust, show_rownames=FALSE,color = color,fontsize_row = 3, legend=TRUE,border_color = NA,main = "Hierarchichal clustering", labels_col = c("WT_Sphere_1","WT_Sphere_2","WT_Sphere_3","SMCHD1_MZ_4_Sphere_1","SMCHD1_MZ_4_Sphere_2", "SMCHD1_MZ_4_Sphere_3"), angle_col = 45)




#k-means
#clustering
set.seed(123)
nclust = 5
results.coef.kmeans =  kmeans(rlg_z, nclust, nstart=50, iter.max=50)
results.coef = rlg_z[order(results.coef.kmeans$cluster),]
indicator = results.coef.kmeans$cluster[order(results.coef.kmeans$cluster)]

#heatmapping
color = c(colorRampPalette(c("dodgerblue2", "white"))(14), colorRampPalette(c("white", "firebrick2"))(14))
breaksList = seq(-thr, thr, length.out = 29)

heat.map <- pheatmap(results.coef, cluster_col=F, breaks=breaksList, cluster_rows=F, show_rownames=FALSE,color = color,fontsize_row = 3, legend=TRUE,border_color = NA,main = "k-means clustering, k=5", labels_col = c("WT_Sphere_1","WT_Sphere_2","WT_Sphere_3","SMCHD1_MZ_4_Sphere_1","SMCHD1_MZ_4_Sphere_2", "SMCHD1_MZ_4_Sphere_3"), angle_col = 45)

#cluster sizes
table(results.coef.kmeans$cluster)

```

After performing a z-score normalization on the expression of significant DEGs across all conditions, calculate the WSS and silhouette width for 2 to 15 clusters, to determine the correct number of clusters for the analysis. Combine results of both statistics, k = 5 is a reasonable choice. Perform the clustering.

Clusters 1, 2, and 5 are up in WT.

Clusters 3 and 4 are up in Smchd1.







### GO:BP enrichment of clusters
```{r}
library(clusterProfiler)
library(org.Dr.eg.db)
library(stringr)
library(ggplot2)
library(svglite)

bp = list()
for (i in 1:nclust) {
  bp[i] = enrichGO(gene          = names(results.coef.kmeans$cluster[results.coef.kmeans$cluster==i]),
                universe      = rownames(nb),
                OrgDb         = org.Dr.eg.db,
                keyType       = 'ENSEMBL',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 1,
                qvalueCutoff  = 0.1,
                readable      = TRUE)
#gene identifiers are ensemble IDs
#orgdb = organism annotation
#ont           = ontology. GO has MF (molecular function), BP (biological process), CC (cellular component).
#universe = all of the genes, after filtering out rRNA/mtRNA and low expression crap
#padj = FDR method
#no pvalue cutoff

}
#ignore the warning messages

#first check if any of the clusters produce no enrichments
head(bp[[1]])
head(bp[[2]])
head(bp[[3]])
head(bp[[4]])
head(bp[[5]])
# 
# 
# dotplot(bp[[1]], title = "GO:BP Enrichment of Cluster 1", font.size = 10) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))
# dotplot(bp[[2]], title = "GO:BP Enrichment of Cluster 2", font.size = 10) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))
# dotplot(bp[[3]], title = "GO:BP Enrichment of Cluster 3", font.size = 10) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))
# dotplot(bp[[4]], title = "GO:BP Enrichment of Cluster 4", font.size = 10) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))
# dotplot(bp[[5]], title = "GO:BP Enrichment of Cluster 5", font.size = 10) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))


```


### GO:MF enrichment of clusters
``` {r}
mf = list()
for (i in 1:nclust) {
  mf[i] = enrichGO(gene          = names(results.coef.kmeans$cluster[results.coef.kmeans$cluster==i]),
                universe      = rownames(nb),
                OrgDb         = org.Dr.eg.db,
                keyType       = 'ENSEMBL',
                ont           = "MF",
                pAdjustMethod = "BH",
                pvalueCutoff  = 1,
                qvalueCutoff  = 0.1,
                readable      = TRUE)
}
# 
head(mf[[1]])
head(mf[[2]])
head(mf[[3]])
head(mf[[4]])
head(mf[[5]])

# 
# dotplot(mf[[1]], title = "GO:MF Enrichment of Cluster 1", font.size = 10) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))
# dotplot(mf[[2]], title = "GO:MF Enrichment of Cluster 2", font.size = 10) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))
# dotplot(mf[[3]], title = "GO:MF Enrichment of Cluster 3", font.size = 10) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))
# dotplot(mf[[4]], title = "GO:MF Enrichment of Cluster 4", font.size = 10) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))
# dotplot(mf[[5]], title = "GO:MF Enrichment of Cluster 5", font.size = 10) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))

```


### KEGG enrichment of clusters
``` {r}
kegg = list()
for (i in 1:nclust) {
  kegg[i] = enrichKEGG(gene          = as.character(ensembl.genes[ensembl.genes$gene_id %in% names(results.coef.kmeans$cluster[results.coef.kmeans$cluster==i])]$entrezgene),
              universe      = as.character(ensembl.genes[ensembl.genes$gene_id %in% rownames(dds)]$entrezgene),
              organism         = "dre",
              qvalueCutoff  = 0.1)

}

# 
head(kegg[[1]])#no enrichment
head(kegg[[2]])#no enrichment
head(kegg[[3]])
head(kegg[[4]])
head(kegg[[5]])#no enrichment

#the top 10 terms in the cluster. gene ratio is the number you observe vs expected, like effect size
# dotplot(kegg[[1]], title = "KEGG Enrichment of Cluster 1", font.size = 10) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))
# 
# dotplot(kegg[[2]], title = "KEGG Enrichment of Cluster 2", font.size = 10) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))
# 
dotplot(kegg[[3]], title = "KEGG Enrichment of Cluster 3", font.size = 10) + scale_y_discrete(labels=function(x) str_wrap(x, width=40)) + scale_color_gradient(high = "#003f5c", low = "#ffa600") + theme(axis.text.y = element_text(size=12), axis.text.x = element_text(size=12))
# 
# dotplot(kegg[[4]], title = "KEGG Enrichment of Cluster 4", font.size = 10) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))

# dotplot(kegg[[5]], title = "KEGG Enrichment of Cluster 5", font.size = 10) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))


```


### Gene set enrichment analysis
``` {r}
library(fgsea)
library(msigdbr)
library(dplyr)
library(gridExtra)

#get gene sets
#msigdbr_species()

#You must select a collection (the category argument). descriptions are on http://www.gsea-msigdb.org/gsea/msigdb/index.jsp
collection = "C2"
subcat = "CP:KEGG"
d_df = msigdbr(species = "Danio rerio", category = collection, subcategory = subcat)
#length(unique(m_df$gs_name))# number of genesets
fgsea_sets = d_df %>% split(x = .$ensembl_gene, f = .$gs_name) #msigdb now supports ensembl! used to be entrezgene

```
For the moment, we are only looking at MSigDB's hallmark gene sets. From MSigDB: "Hallmark gene sets summarize and represent specific well-defined biological states or processes and display coherent expression. These gene sets were generated by a computational methodology based on identifying gene set overlaps and retaining genes that display coordinate expression. The hallmarks reduce noise and redundancy and provide a better delineated biological space for GSEA. We refer to the original overlapping gene sets, from which a hallmark is derived, as its 'founder' sets. Hallmark gene set pages provide links to the corresponding founder sets for deeper follow up."

All gene set descriptions here: http://www.gsea-msigdb.org/gsea/msigdb/index.jsp

**The most relevant collections for this experiment are C3, or external development related gene sets.**

#### SMCHD1 Sphere vs WT Sphere
```{r}
changes = all_cond #recall that this is before we got significant_results, but after we cleaned and ran negbinom/wald. premise of gsea is that the individual expression changes needn't be significant, so this is what we want to search within.
# sum(ensembl.genes$gene_id %in% row.names(changes)) #the number of genes we matched in the annotation
# changes[!(row.names(changes) %in% ensembl.genes$gene_id),] #the genes in the data we failed to match. They simply do not exist in the annotation. Strange.


ensembl_sub = ensembl.genes[ensembl.genes$gene_id %in% row.names(changes), ] #subset annotation according to overlap
changes_sub = changes[row.names(changes) %in% ensembl.genes$gene_id, ] #subset change data according to overlap

changes_sub = changes_sub[order(row.names(changes_sub)),] #ensure genes are in same order as annotation
sum(ensembl_sub$gene_id == row.names(changes_sub)) #are all gene ids identical?

ensembl_sub$log2FoldChange = changes_sub$log2FoldChange #get corresponding foldchanges

#create vector of fold changes to be ranked by lfc
ensembl_sub_lfc = ensembl_sub$log2FoldChange
names(ensembl_sub_lfc) = ensembl_sub$gene_id

#order by lfc
ensembl_sub_lfc = ensembl_sub_lfc[order(ensembl_sub_lfc, decreasing = T)]
#filter for protein coding genes
ensembl_sub_lfc = ensembl_sub_lfc[names(ensembl_sub_lfc) %in% ensembl_sub$gene_id[ensembl_sub$gene_biotype=="protein_coding"]]
#run gsea
fgseaRes <- fgseaMultilevel(pathways = fgsea_sets, 
                  stats = ensembl_sub_lfc,
                  minSize = 1,
                  maxSize = 500,
                  eps = 0)

#plot
fgseaResTidy = fgseaRes %>% as_tibble() %>% arrange(desc(NES))
fgseaResTidy %>% dplyr::select(-leadingEdge, -ES) %>% arrange(padj) 
# ggplot(fgseaResTidy %>% filter(padj < 0.05) %>% head(n= 20), aes(reorder(pathway, NES), NES)) + coord_flip() + geom_col(aes(fill = NES < 2.5)) + labs(x="Term", y="Normalized Enrichment Score", title=paste("GSEA of SMCHD1 Sphere vs WT Sphere, on Ontology Gene Set",collection)) + theme_minimal() + scale_y_discrete(labels=function(x) str_wrap(x, width=40))

#we can also look at a specific term's enrichment
# plotEnrichment(fgsea_sets[["HALLMARK_ALLOGRAFT_REJECTION"]], ensembl_sub_lfc) + labs(title="HALLMARK_ALLOGRAFT_REJECTION")
```

```{r}
#need a separate chunk or it won't knit properly
#another way to visualize gsea
topPathwaysUp <- fgseaRes[ES > 0][head(order(padj), n=10), pathway]
topPathwaysDown <- fgseaRes[ES < 0][head(order(padj), n=10), pathway]
topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
plotGseaTable(fgsea_sets[topPathways], ensembl_sub_lfc, fgseaRes, gseaParam = 0.5, render = T)
```





### Custom GSEA
```{r}
# pathways = gmtPathways("HuGSEA.txt")
lookup = read.csv("human_zebrafish_hcop_fifteen_column.txt", sep = "\t")
# setname = "Hu et al. (2018)"
# 
# length(unique(lookup$zebrafish_ensembl_gene)) #number of unique orthologs in the entire hcop table
# #compare with ensembl: 25592 coding genes in the zebrafish primary assembly.
# 
# paths = list()
# for (i in 1:length(pathways)) {
#   paths[[i]] = lookup[lookup$human_symbol %in% pathways[[i]], 9]
#   paths[[i]] = paths[[i]][nchar(paths[[i]]) == 18]
#   paths[[i]] = unique(paths[[i]])
# }
# 
# names(paths) = names(pathways) #adjust as needed for multiple pathways
# 
# sapply(paths, length)

```

### Volcano plotting specific orthologs
Using a combination of the annotated volcano plot from earlier and custom gsea ortholog lookup method.

```{r}
library(hrbrthemes)

hbox = read.delim("homeobox_hugo.txt")
hbox = hbox[28:66, 2]
#conversion using objects from custom gsea
# s1 = c("HOXA1","HOXA10","HOXA2")
s1 = c(hbox, "SMCHD1")
o1 = lookup[lookup$human_symbol %in% s1, c(5,9,12)]
o1 = o1[nchar(o1$zebrafish_ensembl_gene) == 18, ]


all_cond_d[all_cond_d$name %in% "ENSDARG00000000175", ] #hoxb2a is present in 4 cell.

#S vs WT
d1 = all_cond_d[all_cond_d$name %in% o1$zebrafish_ensembl_gene, ]
ggplot(all_cond_d, aes(x = log2FoldChange, y = -log10(padj), label=name)) + geom_point(alpha = 0.5, color = "dodgerblue") + labs(title = "Volcano, SMCHD1 Sphere vs WT Sphere") + geom_point(data = d1, aes(x = log2FoldChange, y = -log10(padj)), color="black") + geom_text_repel(data = d1, size = 3) + theme_ipsum() + ylim(0,50)

m1 = merge(o1, d1, by.x = 2, by.y = 7)
m1 = m1 %>% distinct(zebrafish_symbol, .keep_all = T)
ggplot(m1, aes(x = log2FoldChange, y = -log10(padj), label=zebrafish_symbol)) + geom_point(color = "black") + labs(title = "Volcano, SMCHD1 Sphere vs WT Sphere") + geom_text_repel(data = m1, size = 3) + theme_ipsum()


ggplot(m1, aes(x = log2FoldChange, y = -log10(padj), label=zebrafish_symbol)) + geom_point(color = "black") + labs(title = "SMCHD1 Sphere vs WT Sphere") + geom_point(data = m1[m1$padj < 0.01,], aes(x = log2FoldChange, y = -log10(padj)), color="firebrick2") + geom_point(data = m1[m1$padj >= 0.01,], aes(x = log2FoldChange, y = -log10(padj)), color="black") + geom_text_repel(data = m1, size = 3) + theme_ipsum()


#original volcano with zfish symbols
d2 = all_cond_d[all_cond_d$name %in% m1$zebrafish_ensembl_gene, ]
d2$name = m1$zebrafish_symbol

ggplot(all_cond_d, aes(x = log2FoldChange, y = -log10(padj), label=name)) + geom_point(alpha = 0.5, color = "dodgerblue") + labs(title = "Volcano, SMCHD1 Sphere vs WT Sphere") + geom_point(data = d2, aes(x = log2FoldChange, y = -log10(padj)), color="firebrick2") + geom_text_repel(data = d2, size = 4) + theme_ipsum() + ylim(c(0,50))


ggplot(all_cond_d, aes(x = log2FoldChange, y = -log10(padj), label=name)) + geom_point(alpha = 0.5, color = "dodgerblue") + labs(title = "Volcano, SMCHD1 Sphere vs WT Sphere") + geom_point(data = d2, aes(x = log2FoldChange, y = -log10(padj)), color="firebrick2") + geom_text_repel(data = d2, size = 4) + theme_ipsum()

# ggsave("volcano_smchd1_sphere_v3.svg", height = 14, width = 14, units = "in")
```


### KEGG pathway expression


#### KEGG enrichment (ORA)

KEGG enrichment on pairwise comparisons to determine pathways.

```{r}
smwt_sig = all_cond[all_cond$padj < 0.05  & abs(all_cond$log2FoldChange) > 1,]
kegg_smwt = enrichKEGG(gene          = as.character(ensembl.genes[ensembl.genes$gene_id %in% rownames(smwt_sig)]$entrezgene),
              universe      = as.character(ensembl.genes[ensembl.genes$gene_id %in% rownames(dds)]$entrezgene),
              organism         = "dre",
              qvalueCutoff  = 0.1)

#dotplot(kegg_smwt, title = "KEGG Enrichment of SMCHD1 Sphere vs WT Sphere", font.size = 10) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))

head(kegg_smwt@result[,c(1,2,6)])



```


Can decide which pathways to investigate based on pairwise enrichment, but also from clusters (if you want a particular expected regulation up/down).

Another approach is choosing pathways that we are interested in apriori and visualizing changes on them, regardless of whether they show up as enriched in the significant DEGs. For c2orf69, we are interested in fatty acid metabolism and related pathways. 



##### Manual pathway selection
```{r, eval = F}
library(pathview)
smwt_all = all_cond$log2FoldChange
names(smwt_all) = rownames(all_cond)


pv.out = pathview(gene.data = smwt_all, gene.idtype = "ENSEMBL", pathway.id = "00230", species = "dre", out.suffix = "SMWT", kegg.native = T, split.group = F, expand.node = F, low = "dodgerblue2", mid = "gray", high = "firebrick2")

```



#### KEGG enrichment (GSEA/FCS)

```{r}
collection = "C2"
subcat = "CP:KEGG"
d_df = msigdbr(species = "Danio rerio", category = collection, subcategory = subcat)
#length(unique(m_df$gs_name))# number of genesets
fgsea_sets = d_df %>% split(x = .$ensembl_gene, f = .$gs_name) 

changes = all_cond

ensembl_sub = ensembl.genes[ensembl.genes$gene_id %in% row.names(changes), ] #subset annotation according to overlap
changes_sub = changes[row.names(changes) %in% ensembl.genes$gene_id, ] #subset change data according to overlap

changes_sub = changes_sub[order(row.names(changes_sub)),] #ensure genes are in same order as annotation
sum(ensembl_sub$gene_id == row.names(changes_sub)) #are all gene ids identical?

ensembl_sub$log2FoldChange = changes_sub$log2FoldChange #get corresponding foldchanges

#create vector of fold changes to be ranked by lfc
ensembl_sub_lfc = ensembl_sub$log2FoldChange
names(ensembl_sub_lfc) = ensembl_sub$gene_id

#order by lfc
ensembl_sub_lfc = ensembl_sub_lfc[order(ensembl_sub_lfc, decreasing = T)]
#filter for protein coding genes
ensembl_sub_lfc = ensembl_sub_lfc[names(ensembl_sub_lfc) %in% ensembl_sub$gene_id[ensembl_sub$gene_biotype=="protein_coding"]]
#run gsea
fgseaRes <- fgseaMultilevel(pathways = fgsea_sets, 
                  stats = ensembl_sub_lfc,
                  minSize = 1,
                  maxSize = 500,
                  eps = 0)

#top 10 most significant by padj
fgseaRes[head(order(padj), n=20), ]
```


```{r}
topPathwaysUp <- fgseaRes[ES > 0][head(order(padj), n=10), pathway]
topPathwaysDown <- fgseaRes[ES < 0][head(order(padj), n=10), pathway]
topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
plotGseaTable(fgsea_sets[topPathways], ensembl_sub_lfc, fgseaRes, gseaParam = 0.5, render = T)
```



##### Automatic pathway selection from GSEA
```{r, eval = F}
library(pathview)
library(KEGGREST)
#also need stringr

smwt_all = all_cond$log2FoldChange
names(smwt_all) = rownames(all_cond)

#list of gsea enrichments
n = fgseaRes[head(order(padj), n=10), 1]
#this is where you could add criteria for selection of gsea pathways (i.e. padj or nes)


for (i in 1:nrow(n)) {
  n1 = n[i]
  n1 = gsub("_", " ", str_sub(n1, 6, nchar(n1)))
  t1 = names(keggFind("pathway", n1)[1])
  t1 = str_sub(t1, 9, nchar(t1))
  pv.out = pathview(gene.data = smwt_all, gene.idtype = "ENSEMBL", pathway.id = t1, species = "dre", out.suffix = "SMWT", kegg.native = T, split.group = F, expand.node = F, low = "dodgerblue2", mid = "gray", high = "firebrick2")
}


```


### GO enrichment of all DEGs


```{r}

# y1 = read.csv("S1.csv", header = T)
# y5 = read.csv("S5.csv", header = T)
# y6 = read.csv("S6.csv", header = T)
# 

fc = as.array(sig_results[,2])
row.names(fc) = row.names(sig_results)


GObp = enrichGO(gene          = rownames(sig_results),
                universe      = rownames(nb),
                OrgDb         = org.Dr.eg.db,
                keyType       = 'ENSEMBL',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 1,
                qvalueCutoff  = 1,
                readable      = TRUE)


GOmf = enrichGO(gene          = rownames(sig_results),
                universe      = rownames(nb),
                OrgDb         = org.Dr.eg.db,
                keyType       = 'ENSEMBL',
                ont           = "MF",
                pAdjustMethod = "BH",
                pvalueCutoff  = 1,
                qvalueCutoff  = 1,
                readable      = TRUE,
                minGSSize = 0)




dotplot(GObp, title = "GO:BP Enrichment of DEGs", font.size = 10) + scale_y_discrete(labels=function(x) str_wrap(x, width=40)) + scale_color_gradient(high = "#003f5c", low = "#ffa600") + theme(axis.text.y = element_text(size=12), axis.text.x = element_text(size=12))





dotplot(GOmf, title = "GO:MF Enrichment of DEGs", font.size = 10) + scale_y_discrete(labels=function(x) str_wrap(x, width=40)) + scale_color_gradient(high = "#003f5c", low = "#ffa600") + theme(axis.text.y = element_text(size=12), axis.text.x = element_text(size=12))










# 
# 
# 
# mfresult = GOmf@result
# 
# vterms = read.csv("vesterlundterms.csv", header = T)
# 
# mfterms = mfresult[mfresult$Description %in% vterms$Term, ]
# 
# write.csv(mfterms, "mfterms_4cell.csv")
# 
# 
# 
# getAnywhere("enricher_internal")
# heatplot(GObp, showCategory = 10, foldChange = fc) + scale_y_discrete(labels=function(x) str_wrap(x, width=20))
# 
# heatplot(GOmf, showCategory = 10, foldChange = fc) + scale_y_discrete(labels=function(x) str_wrap(x, width=20))





```




### END














